File: files/filehatch/src/main/java/example/micronaut/filehatch/FileController.java
package example.micronaut.filehatch;

import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.reactivex.Flowable;

@Controller("/files")
public class FileController {

    private final FilesOperations filesOperations;
    private final FileDirectoryOperations fileDirectoryOperations;

    public FileController(FilesOperations filesOperations,
                          FileDirectoryOperations fileDirectoryOperations) {
        this.filesOperations = filesOperations;
        this.fileDirectoryOperations = fileDirectoryOperations;
    }

    @Get
    public Flowable<FileHatch> index() {
        return filesOperations.findAll()
                .flatMapMaybe(b -> fileDirectoryOperations.stock(b.getIsbn())
                        .filter(Boolean::booleanValue)
                        .map(rsp -> b)
                ).map(file -> new FileHatch(file.getName()));
    }
}

File: files/filehatch/src/main/kotlin/example/micronaut/FileController.kt
package example.micronaut

import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Get
import io.reactivex.Flowable

@Controller("/files") // <1>
class FileController(private val fileOperations: FileOperations,
                     private val fileDirectoryOperations: FileDirectoryOperations) { // <2>

    @Get // <3>
    fun index(): Flowable<FileHatch> {
        return fileOperations.findAll()
                .flatMapMaybe { b ->
                    fileDirectoryOperations.stock(b.isbn)
                            .filter { hasStock -> hasStock }
                            .map { _ -> b }
                }.map { (_, name) -> FileHatch(name) }
    }
}
